### PR TITLE
Source project controller delete action use pundit

### PR DESCRIPTION
**Pull request by vpereira**
instead of custom authorization, use pundit to do authorization

### CODE REVIEW

**Comment by marcus-h** [comment]
I just had a very quick glance at your patch... so the following might
be completely wrong:)

On 2018-08-24 08:22:46 -0700, Victor Pereira wrote:
> diff --git a/src/api/app/controllers/source_project_controller.rb b/src/api/app/controllers/source_project_controller.rb
> index 1c4797267e..bc29773405 100644
> --- a/src/api/app/controllers/source_project_controller.rb
> +++ b/src/api/app/controllers/source_project_controller.rb
> @@ -55,12 +57,7 @@ def delete
>      project = Project.get_by_name(params[:project])
>  
>      # checks
> -    unless project.is_a?(Project) && User.current.can_modify?(project)
> -      logger.debug "No permission to delete project #{project}"
> -      render_error status: 403, errorcode: 'delete_project_no_permission',
> -                   message: "Permission denied (delete project #{project})"
> -      return
> -    end
> +    authorize(project, policy_class: ProjectPolicy)
>      project.check_weak_dependencies!
>      opts = { no_write_to_backend: true,
>               force:               params[:force].present?,
> 
> 
> diff --git a/src/api/app/policies/project_policy.rb b/src/api/app/policies/project_policy.rb
> index 2b0946b588..81f4893c69 100644
> --- a/src/api/app/policies/project_policy.rb
> +++ b/src/api/app/policies/project_policy.rb
> @@ -35,4 +34,13 @@ def can_create_package_in?
>    def local_project_and_allowed_to_create_package_in?
>      local? && can_create_package_in?
>    end
> +
> +  def update_or_delete
> +    return false unless local_project_and_allowed_to_create_package_in?
> +    # The ordering is important because of the lock status check
> +    return true if @user.is_admin?
> +    return false unless @user.can_modify?(@record, true)
> +    # Regular users are not allowed to modify projects with remote references
> +    no_remote_instance_defined_and_has_not_remote_repositories?
> +  end
>  end

Doesn't this break the following scenario?
- create a "home:<user>" project, if non-existent
- remove all <person/> elements from the "home:<user>" project meta
- curl -u <user> -X DELETE https://<apiurl>/source/home:<user>

With the new code the DELETE request is going to fail because
local_project_and_allowed_to_create_package_in? returns false
(since @user.can_create_package_in?(...) returns false).

With the old code the DELETE request is going to succeed because
User.current.can_modify?(project) returns true (due to the special
"home project handling" in can_modify_project_internal).

**Comment by marcus-h** [comment]
On 2018-08-26 11:15:21 -0700, Victor Pereira wrote:
> looking in the https://github.com/openSUSE/open-build-service/blob/master/src/api/app/controllers/webui/project_controller.rb looks like we do exactly the same for the webui. Do you imagine a scenario where the API and the "webui" should work differently here?

No, ideally they do the same:)

> BTW, if i understood correctly - i will test it tomorrow - but the scenario that you described, the user shouldn't be able to delete the project, if he isn't maintainer anymore (looks crazy with home directory, but if the user hasn't permission, just because he created the directory, doesn't mean he should be able to delete it, no?)
> 
Hmm I think it depends on the POV here (that is, if we want to treat a
home project as a special case or not). It might be a bit "inconsistent"
if we allow a user that has no "explicit" permissions in her home:<user>
to create/modify/etc. packages, but disallow the deletion of the project.

The only thing I wanted to point out is that the patch changes the API:)

**Comment by vpereira** [comment]
> Ok... now I'm officially 30/thirsty/somewhat years old...  so now

Alles gute zum Geburtstag!

> everything I write is probably undefined behavior... ;)
>  https://github.com/vpereira/open-build-service/blob/source_project_controller_delete_pundit/src/api/spec/controllers/source_project_controller_spec.rb#L29
> Please bear/beer with me, but I don't buy it:) (except you exactly point
> out my fault:) (which is quite likely) )
> Again, you're probably right, but where's my wrong reasoning?:)
> (I just survived being thirty \o/ (so don't take everything literally;) )

webui and api should have the same authorization rule, and I don't see why if an user isn't a maintainer of a project, even of his own home project, why he should be able to delete it. Imagine that in your operating system? You do remove your own permissions from $HOME, but you still be able to delete it? no way :)

**Comment by adrianschroeter** [comment]
yes, webui and api should behave the same, but please adapt behaviour of webui and try to keep api stable (except there is a reason why the api behaviour is wrong, but it is here a matter of taste IMHO).

**Comment by adrianschroeter** [comment]
Happy birthday Marcus .. thirty .. live is over then. well, make the best of the few remaining days ;)

**Comment by vpereira** [comment]
> yes, webui and api should behave the same, but please adapt behaviour of webui and try to keep api stable (except there is a reason why the api behaviour is wrong, but it is here a matter of taste IMHO).

API still stable, the use case described by @marcus-h wasn't even supported by any test case, and probably different authorization rules between UI and API can lead to confusion or security issues.

**Comment by marcus-h** [comment]
On 2018-08-27 15:35:21 +0000, Victor Pereira wrote:
> @marcus-h I even added a test case to make sure that it cannot happen :) 
> 
> https://github.com/vpereira/open-build-service/blob/source_project_controller_delete_pundit/src/api/spec/controllers/source_project_controller_spec.rb#L29
> 
Ah now I got your comment and testcase right (I was looking at
lines 25 and above). Please ignore my previous reply - sometimes
one shouldn't write mails... ;)

File: src/api/app/policies/project_policy.rb [hunk 1]
```diff
@@ -35,4 +34,13 @@ def can_create_package_in?
   def local_project_and_allowed_to_create_package_in?
     local? && can_create_package_in?
   end
+
+  def update_or_delete
```
Comment (bgeuken) [inline src/api/app/policies/project_policy.rb]: The name sounds like this method would delete or update something. I would rename it to something like `write_permissions?`

**Comment by bgeuken** [comment]
@adrianschroeter IMO it makes sense to permit deletion of projects unless the requester has permissions. Is there any reason to make an exception for home projects?

**Comment by vpereira** [comment]
@adrianschroeter I created this PR https://github.com/openSUSE/open-build-service/pull/5758 that move the existent logic from API to pundit, and added some test cases.. however I still prefer to go with the  option, where no special case for home projects exists.

@bgeuken nice catch, as it is in the webui today, if the user removes all maintainers in his home project, he shoots himself in the foot, needing an admin to put him back. To avoid, IMO we should avoid "top projects" without maintainer, but not completely bypass the authorization

**Comment by hennevogel** [comment]
> where no special case for home projects exists

Why? It should not be possible for people to not change their home project anymore. They should not need to bother an admin. There is **no** need for this.

**Comment by vpereira** [comment]
> Why? It should not be possible for people to not change their home project anymore. They should not need to bother an admin. There is no need for this.

@hennevogel  there are two issues here:

a) webui and api have different authorization rules for the same use case. this is an open issue.
b)  to avoid the user to be removed from the maintainer role, or avoid project _meta without maintainer in top level projects, is IMO, better than silently bypass the authorization.

**Comment by bgeuken** [comment]
IMO, we should not allow to remove all maintainers from a top level project and always create a home project for new users.
But that's out of scope for this PR. So it seems you need to add the exception to the policy to allow users to create their home project. I would check how the webui handles this right now.

**Comment by bgeuken** [comment]
Since you addressed some the discussions here in #5758... shall we close this one or how can we proceed?

**Comment by vpereira** [comment]
@bgeuken we can do it. The authorization based on the render format (xml one rule, html other rule) issue still unresolved though..

**Pull request closed without merging**